python/python/bystro/ancestry/model.py

```diff
@@ -0,0 +1,185 @@
+
+    Raises:
+        ValueError: If the PCA or RFC file is not found.
+
+    Returns:
+        AncestryModel: The loaded ancestry model.
+    """
+    s3_client = boto3.client("s3")
+
+    logger.info("Downloading PCA file %s", pca_file_key)
+
+    with Timer() as timer:
+        try:
+            s3_client.download_file(Bucket=ANCESTRY_BUCKET, Key=pca_file_key, Filename=pca_local_path)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "NoSuchKey":
+                raise ValueError(f"{pca_file_key} not found. This assembly is not supported.") from e
+            raise  # Re-raise the exception if it's not a "NoSuchKey" error
+
+        try:
+            s3_client.download_file(Bucket=ANCESTRY_BUCKET, Key=rfc_file_key, Filename=rfc_local_path)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "NoSuchKey":
+                raise ValueError(
+                    f"{rfc_file_key} ancestry model not found. This assembly is not supported."
+                ) from e
+            raise
+
+    logger.debug("Downloaded PCA file and RFC file in %f seconds", timer.elapsed_time)
+
+    return get_one_model_from_file_system(pca_local_path, rfc_local_path)
+
+
+def get_models_from_s3(assembly: str) -> AncestryModels:
+    """
+    Load the ancestry models for the given assembly from S3.
+
+    Args:
+        assembly (str): The genome assembly to load the models for.
+            Example: "hg38" or "hg19"
+
+    Returns:
+        AncestryModels: The loaded models for the given assembly.
+    """
+    if assembly in models_cache:
+        logger.info("Model for assembly %s found in cache.", assembly)
+        return models_cache[assembly]
+
+    pca_local_key_gnomad = f"{assembly}_{GNOMAD_PCA_FILE}"
+    rfc_local_key_gnomad = f"{assembly}_{GNOMAD_RFC_FILE}"
+
+    pca_file_key_gnomad = f"{assembly}/{pca_local_key_gnomad}"
+    rfc_file_key_gnomad = f"{assembly}/{rfc_local_key_gnomad}"
+
+    pca_local_key_array = f"{assembly}_{ARRAY_PCA_FILE}"
+    rfc_local_key_array = f"{assembly}_{ARRAY_RFC_FILE}"
+
+    pca_file_key_array = f"{assembly}/{pca_local_key_array}"
+    rfc_file_key_array = f"{assembly}/{rfc_local_key_array}"
+
+    gnomad_model = get_one_model_from_s3(
+        pca_local_key_gnomad, rfc_local_key_gnomad, pca_file_key_gnomad, rfc_file_key_gnomad
+    )
+    array_model = get_one_model_from_s3(
+        pca_local_key_array, rfc_local_key_array, pca_file_key_array, rfc_file_key_array
+    )
+
+    models = AncestryModels(gnomad_model, array_model)
+
+    # Update the cache with the new model
+    if len(models_cache) >= 1:
+        # Remove the oldest loaded model to maintain cache size
+        oldest_assembly = next(iter(models_cache))
+        del models_cache[oldest_assembly]
+    models_cache[assembly] = models
+
+    return models
+
+
+def get_one_model_from_file_system(pca_path: str, rfc_path: str) -> AncestryModel:
```

Comment:
the local filesystem versions are needed because our s3 bucket is not publicly accessible, and also because in the future we will cache the models to the local disk to save us costs on s3 (frequently pulling data from s3 is not a good idea from cost standpoint)